dcos/dcos
# [DCOS_OSS-444] Introduce toggle for TLS 1.0 support in Admin Router

## High Level Description

* Change the default TLS configuration.
  By default before, TLS 1.0 was enabled.
  By default with this change, TLS 1.0 is disabled.

* Allow toggling TLS 1.0.
  This is with a configuration variable.

## Related Issues

  - [DCOS_OSS-1819](https://jira.dcos.io/browse/DCOS_OSS-1819) Document that Admin Router does not support TLS 1.0 by default anymore.
  - [DCOS_OSS-444](https://jira.dcos.io/browse/DCOS_OSS-444) Admin Router: disable TLS 1.0 support by default.
- [DCOS-13437](https://jira.mesosphere.com/browse/DCOS-13437) Add DC/OS config option to enable Admin Router serving TLS 1.0 (after disabling by default)

## Checklist for all PR's

  - [X] Included a test which will fail if code is reverted but test is not.
  - [X] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [X] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Lower Level Changes

* Move `gen` tests into a `tests` directory.
  This separates them slightly from implementation files but for this change I chose to do it here in order to have a new file for common `gen` testing utilities.

* Move TLS configuration to a generated file.
  This was in the static `packages/adminrouter/extra/src/common/server.conf` file.
  This is now in the generated `adminrouter-tls.conf` file.
  Only the SSL protocols line needs to be in a generated file.
  However, it seemed sensible to keep all TLS configuration together.

* Add a "realistic" saple configuration for the new config file to the Admin Router test Docker container.

## Further considered work

* Rather than instructing on manual tests, add these to CI.
  That relies on test infrastructure which is not yet built.
  That test infrastructure would associate configurations with tests and create clusters.

* Rather than a textual description of commands to run, create a small script.
  I think that the effort is better spent on the aforementioned test infrastructure.
  The manual steps are very simple.

* Add a test that the Admin Router Docker container has a "realistic" configuration.
  That is, add a test which would fail if the new `packages/adminrouter/docker/adminrouter-tls.conf` sample did not exist.
  I am not sure of the trade-offs involved with doing this.

* The comments referred to in the code are in the private JIRA.
  Perhaps move these and the references to the public JIRA.

I am happy to split this up as a reviewer wishes.

## Manual testing

See comments on https://jira.mesosphere.com/browse/DCOS-13437 for a detailed description of the manual testing and the motivations behind it.

I ran the following commands on the `master` branch and this branch to demonstrate the difference between the two:

 ```
 # I ran the following with tls1, tls1_1 and tls1_2 on master nodes
 dcos-shell openssl s_client -connect localhost:443 -tls1
 
 # I ran the following with tls1, tls1_1 and tls1_2 on agent nodes
 dcos-shell openssl s_client -connect localhost:61002 -tls1
 ```
 
 I considered "Works" to be that the output included "Connected" and a certificate.
 I considered "Not supported" to be that the output included "Connected" and no certificate.

 I got the following results:
 
- Master Branch:
	- Master node:
		- TLS 1.0 - Works
		- TLS 1.1 - Works
		- TLS 1.2 - Works
	- Agent node:
		- TLS 1.0 - Connection Refused
		- TLS 1.1 - Connection Refused
		- TLS 1.2 - Connection Refused
- New branch TLS 1.0 enabled:
	- Master node:
		- TLS 1.0 - Works
		- TLS 1.1 - Works
		- TLS 1.2 - Works
	- Agent node:
		- TLS 1.0 - Connection Refused
		- TLS 1.1 - Connection Refused
		- TLS 1.2 - Connection Refused
- New branch TLS 1.0 disabled:
	- Master node:
		- TLS 1.0 - Not supported
		- TLS 1.1 - Works
		- TLS 1.2 - Works
	- Agent node:
		- TLS 1.0 - Connection Refused
		- TLS 1.1 - Connection Refused
		- TLS 1.2 - Connection Refused

## Other information

Evidence that this does not break DC/OS Enterprise is at https://github.com/mesosphere/dcos-enterprise/pull/803/files.
